runtime/vm/visible.c

```diff
@@ -142,3 +155,55 @@ checkVisibility(J9VMThread *currentThread, J9Class* sourceClass, J9Class* destCl
 
 	return result;
 }
+
+#if defined(J9VM_OPT_VALHALLA_NESTMATES)
+static VMINLINE UDATA loadAndVerifyNestHost(J9VMThread *vmThread, J9Class *clazz) {
+	J9Class *nestHost = NULL;
+	UDATA result = J9_VISIBILITY_ALLOWED;
+	J9ROMClass *romClass = clazz->romClass;
+	J9UTF8 *nestHostName = J9ROMCLASS_NESTHOSTNAME(romClass);
+
+	/* If no nest host is named, class is own nest host */
+	if (NULL == nestHostName) {
+		nestHost = clazz;
+	} else {
+		UDATA classLoadingFlags = J9_FINDCLASS_FLAG_THROW_ON_FAIL;
+		nestHost = vmThread->javaVM->internalVMFunctions->internalFindClassUTF8(vmThread, J9UTF8_DATA(nestHostName), J9UTF8_LENGTH(nestHostName), clazz->classLoader, classLoadingFlags);
+
+		/* Nest host must successfully loaded by the same classloader in the same package & verify the nest member */
+		if (NULL == nestHost) {
+			result = J9_VISIBILITY_NEST_HOST_LOADING_FAILURE_ERROR;
+		} else if (clazz->classLoader != nestHost->classLoader) {
+			result = J9_VISIBILITY_NEST_HOST_DIFFERENT_CLASSLOADER_ERROR;
+		} else if (clazz->packageID != nestHost->packageID) {
+			result = J9_VISIBILITY_NEST_HOST_DIFFERENT_PACKAGE_ERROR;
+		} else {
+			/* The nest host must have a nestmembers attribute that claims this class. */
+			BOOLEAN verified = FALSE;
+			J9UTF8 *className = J9ROMCLASS_CLASSNAME(romClass);
+			J9SRP *nestMembers = J9ROMCLASS_NESTMEMBERS(nestHost->romClass);
+			U_16 nestMemberCount = nestHost->romClass->nestMemberCount;
+			U_16 i = 0;
+
+			for (i = 0; i < nestMemberCount; i++) {
+				J9UTF8 *nestMemberName = NNSRP_GET(nestMembers[i], J9UTF8*);
+				if (J9UTF8_EQUALS(className, nestMemberName)) {
+					verified = TRUE;
+					break;
+				}
+			}
+			if (!verified) {
+				result = J9_VISIBILITY_NEST_MEMBER_NOT_CLAIMED_ERROR;
+			}
+		}
+
+		/* If any problem occurred in nest host verification, mask 0th bit */
+		if (J9_VISIBILITY_ALLOWED != result) {
```

Comment:
do you still need this if you are returning the `J9_VISIBILITY_NEST*` error code?

Reply:
I need some way of identifying which class - destClass or sourceClass - causes the nest verification error. I thought that the best way to do so was flag the 0 bit. A quick overview of the other options that I saw and my reasoning is below. Am open to other ideas/suggestions - having a non-functional non-null pointer may not be a good idea

---

Options:
1. Separate error codes for nest host errors in destClass & sourceClass
2. Changing function headers to pass a boolean/flag/other parameter to indicate nest host errors in either destClass or sourceClass
3. Flagging the 0th bit on the address
4. Putting a non-address value on the nest host (eg. -1)
5. Making error messages non-dependent on which classes have issues

---

Issues:
1. On a conceptual level, nest loading errors in either class would raise the same type of error, but  nothing explicitly wrong with it (other than longer if-statements).
2. Creates excessive changes to existing code.
3. Equivalent to this method, but would require searching for a class if later changes to the code want to access the nest host.
4. Risks that later changes to the code try to use the address to access nest host without masking the 0 bit; may cause confusion later.
5. Need to know which class has nest loading issue, in order to raise accurate errors.

Reply:
As discussed yesterday, new/current approach is:
* Create two sets of visibility error codes; one set for nest visibility/verification issues in the source class and one set for nest visibility/verification issues in the destination class
* Shift the value of these visibility error codes such that they can be separated from the non-nest related error codes with a single comparison and that the source class codes and destination class codes can be separated from eachother via a single bitwise comparison.

eg.

Visibility check result | Value | Hex value | Last eight bits
--- | --- | --- | ---
J9_VISIBILITY_ALLOWED | 1 | 1 | 0000 0001
J9_VISIBILITY_NON_MODULE_ACCESS_ERROR | 0 | 0 | 0
J9_VISIBILITY_MODULE_READ_ACCESS_ERROR | -1 | FFFF FFFF | 1111 1111
J9_VISIBILITY_MODULE_PACKAGE_EXPORT_ERROR | -2 | FFFF FFFE | 1111 1110 
J9_VISIBILITY_NEST_HOST_LOADING_FAILURE_ERROR_SOURCE_CLASS | -17 | FFFF FFFF | 1110 1111
J9_VISIBILITY_NEST_HOST_LOADING_FAILURE_ERROR_TARGET_CLASS | -18 | FFFF FFEF | 1110 1110
J9_VISIBILITY_NEST_HOST_DIFFERENT_PACKAGE_ERROR_SOURCE_CLASS | -19 |  FFFF FFED | 1110 1101
J9_VISIBILITY_NEST_HOST_DIFFERENT_PACKAGE_ERROR_TARGET_CLASS | -20 | FFFF FFEC | 1110 1100
J9_VISIBILITY_NEST_MEMBER_NOT_CLAIMED_ERROR_SOURCE_CLASS | -21 | FFFF FFEB | 1110 1011
J9_VISIBILITY_NEST_MEMBER_NOT_CLAIMED_ERROR_TARGET_CLASS | -22 | FFFF FFEA | 1110 1010

Then we can detect a nest-related error via checking the range or a single bitwise comparison on the 5th- 8th last bits
```
if ((J9_VISIBILITY_NEST_HOST_LOADING_FAILURE_ERROR_SOURCE_CLASS >= errorCode)
    && (J9_VISIBILITY_NEST_MEMBER_NOT_CLAIMED_ERROR_TARGET_CLASS <= errorCode)) { ... }
```
```
#define J9_VISIBILITY_NEST_CODE 240 // 0000 ... 0000 1111 0000
#define J9_VISIBILITY_NEST_MASK 224 // 0000 ... 0000 1110 0000
if (J9_VISIBILITY_NEST_CODE == (J9_VISIBILITY_NEST_MASK & errorCode)) { ... }
```
And we can also use the parity bit to detect if the source class or destination class is causing the error/exception
```
if (0 == errorCode % 2) { ... /* target class */ }
if (1 == errorCode % 2) { ... /* destination class */ }
```